### ISSUE TITLE
UNRESOLVABLE_REFERENCE for circular references

### ISSUE THREAD

**Issue by drej1** [apigee-127/sway#165]
Hi,
In sway you are calling helpers.removeCirculars(results.definitionFullyResolved); while the removeCirculars function assumes the definition is fully resolved but it is not, because json-refs is not called with option resolveCirculars=true and by default it is set to false, so the circulars are detected but not resolved. This way there are no repeating nodes inside removeCirculars function, nothing is replaces with {}.
And UNRESOLVABLE_REFERENCE is thrown by z-schema schema compilation.

Please can you fix sway to call json-refs with resolveCirculars=true option?

Or maybe change the logic of the removeCirculars method to use json-refs's circular=true metadata flag even when the circulars are not resolved?

Thanks.

example swagger.yaml snippet:
```yaml
CategoryModel:
    type: object
    properties:
      _id:
        type: string
        description: category name (id)
      displayName:
       type: string
       description: category display name
      imgUrl:
        type: string
        description: category image url
      parent:
        $ref: "#/definitions/CategoryModel"
      ancestors:
        type: array
        items: 
          $ref: "#/definitions/CategoryModel"
```

**Comment by penx** [comment]
As mentioned in #126, I am having the same issue, and have made the following example to help reproduce:

https://github.com/penx/sway-resolve-reference

**Comment by penx** [comment]
I've used my forks of swagger-node-runner and sway for the example, mainly because:

- I want to ref the latest version of sway/json-refs
- I'm not sure how to use Sway as a mock server on its own without using swagger-node-runner 

The forks shouldn't contain any changes that impact this bug, so I expect you could also replicate the issue just by using the spec file.

**Comment by penx** [comment]
@whitlockjc if you have any pointers to fixing this, I may have some time to look at it

**Comment by penx** [comment]
I've created a failing unit test and then patched with @drej1's suggestion, in PR here:

https://github.com/apigee-127/sway/pull/172

**Comment by whitlockjc** [comment]
Is this still happening?  `json-refs@v3.0.0+` should fix this.

**Comment by whitlockjc** [comment]
Now that `json-refs` support circulars, we'll update `sway` to work with it **but** there is already a way to pass options to `json-refs` allowing you to do this.  [Sway.create](https://github.com/apigee-127/sway/blob/master/docs/API.md#swaycreateoptions-customformats-customformatgenerators--promise) has a `jsonRefs` option that is passed to `json-refs`.

**Comment by penx** [comment]
I won’t have a chance to test for a week or so, but you could try the unit test I wrote in #172 and merge the unit test only if it’s now working

**Comment by whitlockjc** [comment]
So I think this is fixed with the newer versions _(`3.0.x+`)_ of `json-refs`.  With your input document, I don't see any failures when using `Sway.create` and then validating the document.
